File: java/Trees/KthSmallestNumberInBSTRecursive.java
import java.util.ArrayList;
import java.util.List;

import DS.TreeNode;

/*
    // Definition of TreeNode:
    class TreeNode {
        public int val;
        public TreeNode left;
        public TreeNode right;
        public TreeNode(int val) {
            this.val = val;
        }
    }
 */

public class KthSmallestNumberInBSTRecursive {
    public int kthSmallestNumberInBSTRecursive(TreeNode root, int k) {
        List<Integer> sortedList = new ArrayList<>();
        inorder(root, sortedList);
        return sortedList.get(k - 1);
    }

    // Inorder traversal function to attain a sorted list of nodes from the BST.
    private void inorder(TreeNode node, List<Integer> sortedList) {
        if (node == null) {
            return;
        }
        inorder(node.left, sortedList);
        sortedList.add(node.val);
        inorder(node.right, sortedList);
    }
}

File: kotlin/Trees/KthSmallestNumberInBSTRecursive.kt
/*
    Definition of TreeNode:

    data class TreeNode(
        var value: Int,
        var left: TreeNode? = null,
        var right: TreeNode? = null
    )
*/

fun kthSmallestNumberInBSTRecursive(root: TreeNode?, k: Int): Int {
    val sortedList = inorder(root)
    return sortedList[k - 1]
}

// Inorder traversal function to attain a sorted list of nodes from the BST.
fun inorder(node: TreeNode?): List<Int> {
    if (node == null) {
        return emptyList()
    }
    return inorder(node.left) + listOf(node.value) + inorder(node.right)
}
